plonk/src/verifier.rs

```diff
@@ -0,0 +1,217 @@
+            * (proof.bar_b + beta * proof.bar_ssigma_2 + gamma) * alpha * beta * proof.bar_z_w);
+
+        let d_line4 = (proof.t_lo_commit + proof.t_mid_commit.mul(Self::power(&evaluation_challenge, proof.degree.clone() + 1) )
+            + proof.t_hi_commit.mul(Self::power(&evaluation_challenge, proof.degree.clone() * 2 + 2))).mul(z_h_e);
+
+        let d = d_line1 + d_line2 - d_line3 - d_line4;
+
+        #[cfg(test)]
+        println!("Compute [F]");
+
+        let f = d + proof.a_commit.mul(v) + proof.b_commit.mul(v * v) + proof.c_commit.mul(v * v * v)
+            + ssigma_1_c.mul(v * v * v * v) + ssigma_2_c.mul(v * v * v * v * v);
+
+        #[cfg(test)]
+        println!("Compute [E]");
+        let e = -r_0 + v * proof.bar_a + v * v * proof.bar_b + v * v * v * proof.bar_c
+            + v * v * v * v * proof.bar_ssigma_1 + v * v * v * v * v * proof.bar_ssigma_2 + u * proof.bar_z_w;
+        let e = scheme.commit_para(&e);
+
+        #[cfg(test)]
+        println!("Compute left side of paring");
+
+        let pairing_left_side = Bls12_381::pairing(
+            (proof.w_ev_x_commit + proof.w_ev_wx_commit.mul(u)).0,
+            self.srs.get_g2s_ref().clone()
+        );
+
+        #[cfg(test)]
+        println!("Compute right side of paring");
+        let pairing_right_side = Bls12_381::pairing(
+            (proof.w_ev_x_commit.mul(evaluation_challenge) + proof.w_ev_wx_commit.mul(u * evaluation_challenge * w)
+                + f - e).0,
+            self.srs.get_g2_ref().clone()
+        );
+
+        #[cfg(test)]
+        println!("Check pairing");
+
+        assert_eq!(pairing_left_side,pairing_right_side, "Verify: Pairing failed, rejected");
+
+        println!("Accepted!!!");
+
+        return true;
+    }
+
+    fn get_circuit_commitment(&self) -> (KzgCommitment, KzgCommitment, KzgCommitment, KzgCommitment,
+                                         KzgCommitment, KzgCommitment, KzgCommitment, KzgCommitment) {
+
+        let scheme = KzgScheme::new(&self.srs);
+        let q_m_c = scheme.commit(self.gate_constraint.q_mx());
+        let q_l_c = scheme.commit(self.gate_constraint.q_lx());
+        let q_r_c = scheme.commit(self.gate_constraint.q_rx());
+        let q_o_c = scheme.commit(self.gate_constraint.q_ox());
+        let q_c_c = scheme.commit(self.gate_constraint.q_cx());
+        let ssigma1_c = scheme.commit(self.copy_constraint.get_ssigma_1());
+        let ssigma2_c = scheme.commit(self.copy_constraint.get_ssigma_2());
+        let ssigma3_c = scheme.commit(self.copy_constraint.get_ssigma_3());
+
+        (q_m_c, q_l_c, q_r_c, q_o_c, q_c_c, ssigma1_c, ssigma2_c, ssigma3_c)
+    }
+
+    pub fn power(a: &Fr, n: usize) -> Fr {
+        if n == 0 {
+            return Fr::from(1);
+        }
+        let mut cur = Fr::from(1);
+        if n % 2 == 1 {
+            cur = a.clone();
+        }
+
+        let q = Self::power(a, n / 2);
+        let res = q * q * cur;
+        return res;
+    }

```

Comment:
I think the variable named "cur" gave me some wrong impression, i think you should rename it, or write something like this.

```suggestion
    pub fn power(a: &Fr, n: usize) -> Fr {
        if n == 0 {
            return Fr::from(1);
        }

        let q = Self::power(a, n / 2);

        return if n % 2 == 1 {
            q * q * q
        } else {
            q * q
        }
    }
```

Reply:
actually, it's not q*q*q, it's q*q*a.